File: backend/vyos_mappers/protocols/rip.py
"""
RIP Protocol Command Mapper

Handles RIP (Routing Information Protocol) configuration commands.
Supports networks, interfaces, redistribution, and timers.
"""

from typing import List, Dict, Any
from ..base import BaseFeatureMapper


class RIPMapper(BaseFeatureMapper):
    """RIP protocol mapper with all RIP configuration operations"""

    def __init__(self, version: str):
        """Initialize with VyOS version."""
        super().__init__(version)

    # ========================================================================
    # Basic RIP Configuration Paths
    # ========================================================================

    def get_rip(self) -> List[str]:
        """Get command path for creating RIP."""
        return ["protocols", "rip"]

    def get_default_distance(self, distance: str) -> List[str]:
        """Get command path for setting default administrative distance."""
        return ["protocols", "rip", "default-distance", distance]

    def get_default_distance_path(self) -> List[str]:
        """Get command path for default distance (for deletion)."""
        return ["protocols", "rip", "default-distance"]

    def get_version(self, version: str) -> List[str]:
        """Get command path for RIP version (1, 2)."""
        return ["protocols", "rip", "version", version]

    def get_version_path(self) -> List[str]:
        """Get command path for version (for deletion)."""
        return ["protocols", "rip", "version"]

    # ========================================================================
    # Network Configuration Paths
    # ========================================================================

    def get_network(self, network: str) -> List[str]:
        """Get command path for adding a network."""
        return ["protocols", "rip", "network", network]

    # ========================================================================
    # Interface Configuration Paths
    # ========================================================================

    def get_interface(self, interface: str) -> List[str]:
        """Get command path for interface configuration."""
        return ["protocols", "rip", "interface", interface]

    def get_interface_send(self, interface: str, version: str) -> List[str]:
        """Get command path for interface send version."""
        return ["protocols", "rip", "interface", interface, "send", version]

    def get_interface_receive(self, interface: str, version: str) -> List[str]:
        """Get command path for interface receive version."""
        return ["protocols", "rip", "interface", interface, "receive", version]

    def get_interface_split_horizon(self, interface: str) -> List[str]:
        """Get command path for split-horizon."""
        return ["protocols", "rip", "interface", interface, "split-horizon"]

    def get_interface_split_horizon_disable(self, interface: str) -> List[str]:
        """Get command path for disabling split-horizon."""
        return ["protocols", "rip", "interface", interface, "split-horizon", "disable"]

    def get_interface_split_horizon_poison_reverse(self, interface: str) -> List[str]:
        """Get command path for split-horizon poison-reverse."""
        return ["protocols", "rip", "interface", interface, "split-horizon", "poison-reverse"]

    def get_interface_authentication_md5(self, interface: str, key_id: str, md5_key: str) -> List[str]:
        """Get command path for MD5 authentication."""
        return ["protocols", "rip", "interface", interface, "authentication", "md5", key_id, "md5-key", md5_key]

    def get_interface_authentication_plaintext(self, interface: str, password: str) -> List[str]:
        """Get command path for plaintext authentication."""
        return ["protocols", "rip", "interface", interface, "authentication", "plaintext-password", password]

    # ========================================================================
    # Passive Interface Configuration
    # ========================================================================

    def get_passive_interface(self, interface: str) -> List[str]:
        """Get command path for passive interface."""
        return ["protocols", "rip", "passive-interface", interface]

    def get_passive_interface_default(self) -> List[str]:
        """Get command path for passive interface default."""
        return ["protocols", "rip", "passive-interface", "default"]

    # ========================================================================
    # Neighbor Configuration Paths
    # ========================================================================

    def get_neighbor(self, neighbor: str) -> List[str]:
        """Get command path for neighbor (for non-broadcast networks)."""
        return ["protocols", "rip", "neighbor", neighbor]

    # ========================================================================
    # Redistribution Configuration Paths
    # ========================================================================

    def get_redistribute(self, protocol: str) -> List[str]:
        """Get command path for redistributing a protocol."""
        return ["protocols", "rip", "redistribute", protocol]

    def get_redistribute_route_map(self, protocol: str, route_map: str) -> List[str]:
        """Get command path for redistribution route-map."""
        return ["protocols", "rip", "redistribute", protocol, "route-map", route_map]

    def get_redistribute_metric(self, protocol: str, metric: str) -> List[str]:
        """Get command path for redistribution metric."""
        return ["protocols", "rip", "redistribute", protocol, "metric", metric]

    # ========================================================================
    # Default Information Configuration
    # ========================================================================

    def get_default_information_originate(self) -> List[str]:
        """Get command path for originating default information."""
        return ["protocols", "rip", "default-information", "originate"]

    # ========================================================================
    # Distribute List Configuration
    # ========================================================================

    def get_distribute_list_access_list_in(self, access_list: str) -> List[str]:
        """Get command path for distribute list access-list in."""
        return ["protocols", "rip", "distribute-list", "access-list", "in", access_list]

    def get_distribute_list_access_list_out(self, access_list: str) -> List[str]:
        """Get command path for distribute list access-list out."""
        return ["protocols", "rip", "distribute-list", "access-list", "out", access_list]

    def get_distribute_list_interface_access_list_in(self, interface: str, access_list: str) -> List[str]:
        """Get command path for distribute list interface access-list in."""
        return ["protocols", "rip", "distribute-list", "interface", interface, "access-list", "in", access_list]

    def get_distribute_list_interface_access_list_out(self, interface: str, access_list: str) -> List[str]:
        """Get command path for distribute list interface access-list out."""
        return ["protocols", "rip", "distribute-list", "interface", interface, "access-list", "out", access_list]

    def get_distribute_list_prefix_list_in(self, prefix_list: str) -> List[str]:
        """Get command path for distribute list prefix-list in."""
        return ["protocols", "rip", "distribute-list", "prefix-list", "in", prefix_list]

    def get_distribute_list_prefix_list_out(self, prefix_list: str) -> List[str]:
        """Get command path for distribute list prefix-list out."""
        return ["protocols", "rip", "distribute-list", "prefix-list", "out", prefix_list]

    # ========================================================================
    # Timer Configuration
    # ========================================================================

    def get_timers_update(self, seconds: str) -> List[str]:
        """Get command path for update timer."""
        return ["protocols", "rip", "timers", "update", seconds]

    def get_timers_timeout(self, seconds: str) -> List[str]:
        """Get command path for timeout timer."""
        return ["protocols", "rip", "timers", "timeout", seconds]

    def get_timers_garbage_collection(self, seconds: str) -> List[str]:
        """Get command path for garbage collection timer."""
        return ["protocols", "rip", "timers", "garbage-collection", seconds]

    # ========================================================================
    # Config Parsing Methods (for READ operations)
    # ========================================================================

    def parse_rip_config(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Parse full RIP configuration from VyOS.

        Args:
            config: Raw RIP config dictionary from VyOS (under protocols > rip)

        Returns:
            Parsed RIP configuration as dictionary
        """
        if not config:
            return {
                "configured": False,
                "networks": [],
                "interfaces": [],
                "passive_interfaces": [],
                "neighbors": [],
                "redistributions": [],
                "version": None,
                "default_distance": None,
                "default_information_originate": False,
                "timers": None,
            }

        # Parse networks
        networks = []
        network_cfg = config.get("network", [])
        if isinstance(network_cfg, list):
            networks = network_cfg
        elif isinstance(network_cfg, dict):
            networks = list(network_cfg.keys())

        # Parse interfaces
        interfaces = self._parse_interfaces(config.get("interface", {}))

        # Parse passive interfaces
        passive_interfaces = self._parse_passive_interfaces(config.get("passive-interface", {}))

        # Parse neighbors
        neighbors = []
        neighbor_cfg = config.get("neighbor", [])
        if isinstance(neighbor_cfg, list):
            neighbors = neighbor_cfg
        elif isinstance(neighbor_cfg, dict):
            neighbors = list(neighbor_cfg.keys())

        # Parse redistribution
        redistributions = self._parse_redistributions(config.get("redistribute", {}))

        # Parse timers
        timers = self._parse_timers(config.get("timers", {}))

        # Parse default information
        default_info = config.get("default-information", {})
        default_information_originate = "originate" in default_info if isinstance(default_info, dict) else False

        return {
            "configured": True,
            "networks": networks,
            "interfaces": interfaces,
            "passive_interfaces": passive_interfaces,
            "neighbors": neighbors,
            "redistributions": redistributions,
            "version": config.get("version"),
            "default_distance": config.get("default-distance"),
            "default_information_originate": default_information_originate,
            "timers": timers,
        }

    def _parse_interfaces(self, interfaces_config: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse interface configurations."""
        interfaces = []
        for iface_name, iface_cfg in interfaces_config.items():
            if not isinstance(iface_cfg, dict):
                interfaces.append({"name": iface_name})
                continue

            # Parse split-horizon
            split_horizon = None
            split_horizon_cfg = iface_cfg.get("split-horizon", {})
            if isinstance(split_horizon_cfg, dict):
                if "disable" in split_horizon_cfg:
                    split_horizon = "disabled"
                elif "poison-reverse" in split_horizon_cfg:
                    split_horizon = "poison-reverse"
            elif split_horizon_cfg:
                split_horizon = "enabled"

            # Parse authentication
            auth = iface_cfg.get("authentication", {})
            authentication = None
            if auth:
                if "md5" in auth:
                    authentication = {"type": "md5"}
                elif "plaintext-password" in auth:
                    authentication = {"type": "plaintext"}

            interfaces.append({
                "name": iface_name,
                "send": iface_cfg.get("send"),
                "receive": iface_cfg.get("receive"),
                "split_horizon": split_horizon,
                "authentication": authentication,
            })

        return interfaces

    def _parse_passive_interfaces(self, passive_config: Any) -> Dict[str, Any]:
        """Parse passive interface configurations."""
        if isinstance(passive_config, dict):
            return {
                "default": "default" in passive_config,
                "interfaces": [k for k in passive_config.keys() if k != "default"],
            }
        return {"default": False, "interfaces": []}

    def _parse_redistributions(self, redist_config: Dict[str, Any]) -> List[Dict[str, Any]]:
        """Parse redistribution configurations."""
        redistributions = []
        for protocol, protocol_cfg in redist_config.items():
            if isinstance(protocol_cfg, dict):
                redistributions.append({
                    "protocol": protocol,
                    "route_map": protocol_cfg.get("route-map"),
                    "metric": protocol_cfg.get("metric"),
                })
            else:
                redistributions.append({
                    "protocol": protocol,
                    "route_map": None,
                    "metric": None,
                })

        return redistributions

    def _parse_timers(self, timers_config: Dict[str, Any]) -> Dict[str, Any] | None:
        """Parse timer configurations."""
        if not timers_config:
            return None

        return {
            "update": timers_config.get("update"),
            "timeout": timers_config.get("timeout"),
            "garbage_collection": timers_config.get("garbage-collection"),
        }

File: frontend/src/lib/api/rip.ts
/**
 * RIP Protocol API Service
 * Handles all RIP (Routing Information Protocol) related API operations
 */

import { apiClient } from "./client";

// ============================================================================
// Types
// ============================================================================

export interface RIPInterface {
  name: string;
  send?: string | null;
  receive?: string | null;
  split_horizon?: string | null;
  authentication?: { type: string } | null;
}

export interface RIPRedistribution {
  protocol: string;
  route_map?: string | null;
  metric?: number | null;
}

export interface RIPTimers {
  update?: number | null;
  timeout?: number | null;
  garbage_collection?: number | null;
}

export interface RIPPassiveInterfaces {
  default: boolean;
  interfaces: string[];
}

export interface RIPConfig {
  configured: boolean;
  networks: string[];
  interfaces: RIPInterface[];
  passive_interfaces: RIPPassiveInterfaces;
  neighbors: string[];
  redistributions: RIPRedistribution[];
  version?: string | null;
  default_distance?: number | null;
  default_information_originate: boolean;
  timers?: RIPTimers | null;
}

export interface RIPCapabilities {
  versions: { value: string; label: string }[];
  split_horizon_modes: { value: string; label: string; description: string }[];
  redistribute_protocols: { value: string; label: string }[];
  default_timers: {
    update: number;
    timeout: number;
    garbage_collection: number;
  };
}

export interface ApiResponse {
  success: boolean;
  message: string;
  error?: string;
}

// ============================================================================
// Service
// ============================================================================

class RIPService {
  /**
   * Get RIP configuration
   */
  async getConfig(): Promise<RIPConfig> {
    return apiClient.get<RIPConfig>("/vyos/protocols/rip/config");
  }

  /**
   * Get RIP capabilities for the connected VyOS version
   */
  getCapabilities(): RIPCapabilities {
    // Static capabilities for RIP
    return {
      versions: [
        { value: "1", label: "Version 1" },
        { value: "2", label: "Version 2" },
      ],
      split_horizon_modes: [
        { value: "enabled", label: "Enabled", description: "Default split-horizon" },
        { value: "disabled", label: "Disabled", description: "Disable split-horizon" },
        { value: "poison-reverse", label: "Poison Reverse", description: "Split-horizon with poison reverse" },
      ],
      redistribute_protocols: [
        { value: "bgp", label: "BGP" },
        { value: "connected", label: "Connected" },
        { value: "isis", label: "IS-IS" },
        { value: "kernel", label: "Kernel" },
        { value: "ospf", label: "OSPF" },
        { value: "static", label: "Static" },
      ],
      default_timers: {
        update: 30,
        timeout: 180,
        garbage_collection: 120,
      },
    };
  }

  /**
   * Enable RIP with basic configuration
   */
  async enable(options?: {
    version?: string;
    defaultDistance?: number;
    defaultInformationOriginate?: boolean;
  }): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/rip/enable", {
      version: options?.version,
      default_distance: options?.defaultDistance,
      default_information_originate: options?.defaultInformationOriginate ?? false,
    });
  }

  /**
   * Disable RIP completely
   */
  async disable(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/rip/disable");
  }

  // =========================================================================
  // Network operations
  // =========================================================================

  /**
   * Add a network to RIP
   */
  async addNetwork(network: string): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/rip/network", { network });
  }

  /**
   * Remove a network from RIP
   */
  async removeNetwork(network: string): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>(`/vyos/protocols/rip/network/${encodeURIComponent(network)}`);
  }

  // =========================================================================
  // Interface operations
  // =========================================================================

  /**
   * Configure an interface for RIP
   */
  async configureInterface(
    interfaceName: string,
    options?: {
      send?: string;
      receive?: string;
      splitHorizon?: string;
    }
  ): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/rip/interface", {
      interface: interfaceName,
      send: options?.send,
      receive: options?.receive,
      split_horizon: options?.splitHorizon,
    });
  }

  /**
   * Remove interface from RIP
   */
  async removeInterface(interfaceName: string): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>(`/vyos/protocols/rip/interface/${encodeURIComponent(interfaceName)}`);
  }

  // =========================================================================
  // Passive interface operations
  // =========================================================================

  /**
   * Set passive interfaces
   */
  async setPassiveInterfaces(options: {
    default?: boolean;
    interfaces?: string[];
  }): Promise<ApiResponse> {
    return apiClient.put<ApiResponse>("/vyos/protocols/rip/passive-interface", {
      default: options.default ?? false,
      interfaces: options.interfaces ?? [],
    });
  }

  // =========================================================================
  // Neighbor operations
  // =========================================================================

  /**
   * Add a neighbor (for non-broadcast networks)
   */
  async addNeighbor(neighbor: string): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/rip/neighbor", { neighbor });
  }

  /**
   * Remove a neighbor
   */
  async removeNeighbor(neighbor: string): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>(`/vyos/protocols/rip/neighbor/${encodeURIComponent(neighbor)}`);
  }

  // =========================================================================
  // Redistribution operations
  // =========================================================================

  /**
   * Add protocol redistribution
   */
  async addRedistribution(
    protocol: string,
    options?: {
      routeMap?: string;
      metric?: number;
    }
  ): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/rip/redistribute", {
      protocol,
      route_map: options?.routeMap,
      metric: options?.metric,
    });
  }

  /**
   * Remove protocol redistribution
   */
  async removeRedistribution(protocol: string): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>(`/vyos/protocols/rip/redistribute/${encodeURIComponent(protocol)}`);
  }

  // =========================================================================
  // Timer operations
  // =========================================================================

  /**
   * Set RIP timers
   */
  async setTimers(timers: {
    update?: number;
    timeout?: number;
    garbageCollection?: number;
  }): Promise<ApiResponse> {
    return apiClient.put<ApiResponse>("/vyos/protocols/rip/timers", {
      update: timers.update,
      timeout: timers.timeout,
      garbage_collection: timers.garbageCollection,
    });
  }

  /**
   * Reset timers to defaults
   */
  async resetTimers(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/rip/timers");
  }

  // =========================================================================
  // Default information operations
  // =========================================================================

  /**
   * Enable default information originate
   */
  async enableDefaultOriginate(): Promise<ApiResponse> {
    return apiClient.post<ApiResponse>("/vyos/protocols/rip/default-information/originate", {});
  }

  /**
   * Disable default information originate
   */
  async disableDefaultOriginate(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/rip/default-information/originate");
  }

  // =========================================================================
  // Version operations
  // =========================================================================

  /**
   * Set RIP version
   */
  async setVersion(version: string): Promise<ApiResponse> {
    return apiClient.put<ApiResponse>(`/vyos/protocols/rip/version/${version}`, {});
  }

  /**
   * Reset version to default
   */
  async resetVersion(): Promise<ApiResponse> {
    return apiClient.delete<ApiResponse>("/vyos/protocols/rip/version");
  }
}

export const ripService = new RIPService();
